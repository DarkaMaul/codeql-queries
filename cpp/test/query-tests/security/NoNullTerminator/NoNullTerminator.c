#include "../../../include/libc/string_stubs.h"

#define LABEL_BUF_SIZE 8

void good_initializations() {
    // These are correct C-strings
    char xx[3] = "ab";
    printf("%s", xx);
    unsigned char yy[3] = "";
    printf("%s", yy);
    unsigned char aa[3] = "AB\x00";
    printf("%s", aa);
    unsigned char ab[3] = "A\000\000";
    printf("%s", ab);
    unsigned char ac[3] = "\000\000";
    printf("%s", ac);
    unsigned char ad[3] = "\000\000\x00";
    printf("%s", ad);
    wchar_t qq[9] = L"12345678";
    wprintf(L"%ls", qq);
}

// TODO: these are false positives that should not be identified as bug
void good_initializations_explicit_nullbyte() {
    // These are correct C-strings
    char xx[3] = "ab\0";
    printf("%s", xx);
    unsigned char yy[4] = "12\x00""4";
    printf("%s", yy);
    unsigned char aa[3] = "A\000B";
    printf("%s", aa);
    unsigned char ab[8] = "1234\x00""678";
    printf("%s", ab);
    wchar_t qq[8] = L"1234\x00" L"678";
    wprintf(L"%ls", qq);
}

int not_null_terminated_initializations() {
    // These are buffers without null terminators, not C-strings (but "C-string likes")
    char ww[4] = "cdef"; // no warning from compilator
    unsigned char qq[6] = "ghijklllllllll\000"; // warning from compilator
    unsigned char kk[3] = "iop\000\x00"; // warning from compilator
    // Out of bound reads
    printf("%s", ww);
    printf("%s", qq);
    printf("%s", kk);
    return strlen(ww);
}

void not_null_terminated_initializations_widechars() {
    // wide char OOB
    wchar_t secret[] = L"secret";
    wchar_t tt[8] = L"12345678";
    wprintf(L"%ls", tt);

    wchar_t dst[50];
    wcscpy(dst, tt);
    wprintf(L"%ls", dst);
}

void not_a_string() {
    char s[4] = { 'a', 'b', 'c', 'd' }; // valid, but not recommended initialization of not-a-string
    printf("%d", s);
}

int main() {
    good_initializations();
    good_initializations_explicit_nullbyte();
    not_null_terminated_initializations();
    not_null_terminated_initializations_widechars();
    not_a_string();
}

