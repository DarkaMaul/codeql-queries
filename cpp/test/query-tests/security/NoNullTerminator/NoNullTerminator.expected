| NoNullTerminator.c:28:19:28:20 | yy | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (4 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:29:5:29:10 | call to printf | call to printf |
| NoNullTerminator.c:30:19:30:20 | aa | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (3 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:31:5:31:10 | call to printf | call to printf |
| NoNullTerminator.c:32:19:32:20 | ab | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (8 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:33:5:33:10 | call to printf | call to printf |
| NoNullTerminator.c:34:13:34:14 | qq | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (32 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:35:5:35:11 | call to wprintf | call to wprintf |
| NoNullTerminator.c:40:10:40:11 | ww | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (4 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:44:5:44:10 | call to printf | call to printf |
| NoNullTerminator.c:40:10:40:11 | ww | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (4 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:47:12:47:17 | call to strlen | call to strlen |
| NoNullTerminator.c:41:19:41:20 | qq | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (6 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:45:5:45:10 | call to printf | call to printf |
| NoNullTerminator.c:42:19:42:20 | kk | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (3 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:46:5:46:10 | call to printf | call to printf |
| NoNullTerminator.c:53:13:53:14 | tt | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (32 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:54:5:54:11 | call to wprintf | call to wprintf |
| NoNullTerminator.c:53:13:53:14 | tt | String is initialized without terminating null byte, as the arary size is equal to initialization string's length (32 bytes).. String is then used as an argument to a function requiring null-terminated strings: $@ | NoNullTerminator.c:57:5:57:10 | call to wcscpy | call to wcscpy |
